### DISCUSSION TITLE
How to improve XKCP

### DISCUSSION THREAD

**Discussion by gvanas** [XKCP/XKCP#113] [Ideas]
Dear all,

In this post, I wanted to share some ideas of improvements in XKCP. If you have feedback or more suggestions, please do not hesitate to post them here.

Also, speaking for myself, I feel my focus is more on cryptographic research than on implementations, and consequently the time I can dedicate to XKCP is much less than what would probably be needed. **So, any help is more than welcome here!**

- Clean-up
    - Remove Keyak and Ketje as these are obsoleted by Xoofff and Xoodyak
    - Remove the smaller Keccak-p[200], [400] and [800]
    - Replace the state size / state alignment defines with proper state struct types
    - Move the standalone implementations to a separate repository
    - Move the Build system to a submodule (as it is shared with [K12](https://github.com/XKCP/K12))
- Process
    - Improve the documentation
    - Add continuous integration
    - Review or test for security vulnerabilities such as [this one](https://eprint.iacr.org/2023/331)
- New features
    - Unify the x86_64 implementations so that the most appropriate implementation (e.g., SSSE3, AVX2, AVX512) is chosen at run time (as already done in [K12](https://github.com/XKCP/K12))
    - Make easy-to-use packages for `libXKCP`

In another dimension, it would be nice to rewrite XKCP in Rust.

What do you think?

Kind regards,
Gilles

**Comment by chfast** [comment]
Having an easy to build/package `libXKCP` with runtime implementation selection would be really great.

**Reply by gvanas** [reply]
Thanks for your feedback. I agree with this being a priority.

**Reply by gvanas** [reply]
I introduced runtime implementation selection for x86-64 in [this commit](https://github.com/XKCP/XKCP/commit/e7a08f7baa3d43d64f5c21e641cb18fe292f2b75), and more recently @phongn extended it to ARM platforms and improved it for libraries. For more details, see in particular [this PR](https://github.com/XKCP/XKCP/pull/166).

**Comment by konsolebox** [comment]
Hello, regarding "Move the Build system to a submodule K12", I'm not sure what it means but as long as the build system can still be easily parsed as how it is now, I'm happy with it.

I rely on a script I made named [import-xkcp-files.rake](https://github.com/konsolebox/digest-kangarootwelve-ruby/blob/master/rakelib/import-xkcp-files.rake) to extract the files the gem only needs so gem doesn't ship with unnecessary files.

I never had issues with it besides failing to include an asm dependency which was already fixed.

**Comment by konsolebox** [comment]
Wow XKCP/K12 actually existed? I totally missed that for 6 years.
